// Copyright 2022 Flower Labs GmbH. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ==============================================================================

syntax = "proto3";

package flwr.proto;

import "flwr/proto/heartbeat.proto";
import "flwr/proto/log.proto";
import "flwr/proto/node.proto";
import "flwr/proto/message.proto";
import "flwr/proto/run.proto";
import "flwr/proto/fab.proto";
import "flwr/proto/appio.proto";

service ServerAppIo {
  // Return a set of nodes
  rpc GetNodes(GetNodesRequest) returns (GetNodesResponse) {}

  // Create one or more messages
  rpc PushMessages(PushAppMessagesRequest) returns (PushAppMessagesResponse) {}

  // Get message results
  rpc PullMessages(PullAppMessagesRequest) returns (PullAppMessagesResponse) {}

  // Get run details
  rpc GetRun(GetRunRequest) returns (GetRunResponse) {}

  // Get FAB
  rpc GetFab(GetFabRequest) returns (GetFabResponse) {}

  // Pull ServerApp inputs
  rpc PullAppInputs(PullAppInputsRequest) returns (PullAppInputsResponse) {}

  // Push ServerApp outputs
  rpc PushAppOutputs(PushAppOutputsRequest) returns (PushAppOutputsResponse) {}

  // Update the status of a given run
  rpc UpdateRunStatus(UpdateRunStatusRequest)
      returns (UpdateRunStatusResponse) {}

  // Get the status of a given run
  rpc GetRunStatus(GetRunStatusRequest) returns (GetRunStatusResponse) {}

  // Push ServerApp logs
  rpc PushLogs(PushLogsRequest) returns (PushLogsResponse) {}

  // Heartbeat
  rpc SendAppHeartbeat(SendAppHeartbeatRequest)
      returns (SendAppHeartbeatResponse) {}

  // Push Object
  rpc PushObject(PushObjectRequest) returns (PushObjectResponse) {}

  // Pull Object
  rpc PullObject(PullObjectRequest) returns (PullObjectResponse) {}

  // Confirm Message Received
  rpc ConfirmMessageReceived(ConfirmMessageReceivedRequest)
      returns (ConfirmMessageReceivedResponse) {}
}

// GetNodes messages
message GetNodesRequest { uint64 run_id = 1; }
message GetNodesResponse { repeated Node nodes = 1; }
